README.md
# OmegaT round theme plugin

This is OmegaT plugin to provide themes for OmegaT 5.6.0 and later.

## Themes

### Flat Rounded Dark theme

Flat design dark theme that has rounded buttons, boxes and bullets.

### Flat Rounded Light theme

Flat design theme that has rounded buttons, boxes and bullets.

## Dependency

OmegaT round theme plugin depends on [FlatLaf -Flat Look and Feel](https://github.com/JFormDesigner/FlatLaf)
by [FormDev Software](https://www.formdev.com/flatlaf/).

## License

The plugin is distributed under GNU General Public License 3 or later.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [Unreleased]

## [v0.1.0]

### Added
* Set application color style

### Changed
* Automate release with CI
* Bump versions
  * gradle omegat plugin@v1.5.7
* Use git tag for release versioning


[Unreleased]: https://github.com/miurahr/omegat-round-theme/compare/v0.1.0...HEAD
[v0.0.1]: https://github.com/miurahr/omegat-round-theme/compare/v0.0.1...v0.1.0


src/main/java/tokyo/northside/omegat/theme/FlatThemes.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2021 Hiroshi Miura
               Home page: http://www.omegat.org/

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/

package tokyo.northside.omegat.theme;

import javax.swing.UIManager;

public class FlatThemes {

    private FlatThemes() {
    }

    public static void loadPlugins() {
        UIManager.installLookAndFeel("Flat Round Dark", "tokyo.northside.omegat.theme.FlatRoundDarkLaf");
        UIManager.installLookAndFeel("Flat Round Light", "tokyo.northside.omegat.theme.FlatRoundLightLaf");
    }

    public static void unloadPlugins() {
    }
}


src/main/java/tokyo/northside/omegat/theme/RoundThemes.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2021 Hiroshi Miura
               Home page: http://www.omegat.org/

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/

package tokyo.northside.omegat.theme;

import javax.swing.UIManager;

/**
 * OmegaT driver module to register themes.
 */
public final class RoundThemes {

    private RoundThemes() {
    }

    /**
     * Register the plugin.
     */
    public static void loadPlugins() {
        UIManager.installLookAndFeel("Flat Round Dark",
                "tokyo.northside.omegat.theme.FlatRoundDarkLaf");
        UIManager.installLookAndFeel("Flat Round Light",
                "tokyo.northside.omegat.theme.FlatRoundLightLaf");
    }

    /**
     * Unload plugins.
     * Currently do nothing.
     */
    public static void unloadPlugins() {
    }
}


src/main/java/tokyo/northside/omegat/laf/CustomLookAndFeel.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2021 Hiroshi Miura
               Home page: http://www.omegat.org/

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/

package tokyo.northside.omegat.laf;

import org.omegat.util.gui.UIDesignManager;
import org.omegat.gui.theme.IThemeInitializer;

import javax.swing.UIManager;
import java.util.Map;


public class CustomLookAndFeel {

    private CustomLookAndFeel() {
    }

    public static void loadPlugins() {
        UIDesignManager.registerTheme(new CustomTheme("FlatLaf Dark","com.formdev.flatlaf.FlatDarkLaf"));
        UIDesignManager.registerTheme(new CustomTheme("FlatLaf Light","com.formdev.flatlaf.FlatLightLaf"));
        UIDesignManager.registerTheme(new FlatRoundDarkTheme());
        UIDesignManager.registerTheme(new FlatRoundLightTheme());
    }

    public static void unloadPlugins() {
    }

    public static class CustomTheme extends UIManager.LookAndFeelInfo implements IThemeInitializer {
        public CustomTheme(String name, String className) {
            super(name, className);
        }
        public void setup() {}
    }

    public static final class FlatRoundDarkTheme extends CustomTheme implements IThemeInitializer {

        private static final String NAME = "Flat Round Dark";
        private static final String CLASS_NAME = "com.formdev.flatlaf.FlatDarkLaf";

        public FlatRoundDarkTheme() {
            super(NAME, CLASS_NAME);
        }

        @Override
        public void setup() {
            UIManager.put( "Component.arrowType", "triangle" );
            UIManager.put( "Button.arc", 999 );
            UIManager.put( "Component.arc", 999 );
            UIManager.put( "ProgressBar.arc", 999 );
            UIManager.put( "TextComponent.arc", 999 );
        }
    }

    public static final class FlatRoundLightTheme extends CustomTheme implements IThemeInitializer {

        private static final String NAME = "Flat Round Light";
        private static final String CLASS_NAME = "com.formdev.flatlaf.FlatLightLaf";

        public FlatRoundLightTheme() {
            super(NAME, CLASS_NAME);
        }

        @Override
        public void setup() {
            UIManager.put( "Component.arrowType", "triangle" );
            UIManager.put( "Button.arc", 999 );
            UIManager.put( "Component.arc", 999 );
            UIManager.put( "ProgressBar.arc", 999 );
            UIManager.put( "TextComponent.arc", 999 );
        }

        public Map<String, Object> getConfig() {
            return null;
        }
    }
}


src/main/java/tokyo/northside/omegat/theme/FlatRoundLightLaf.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2021 Hiroshi Miura
               Home page: http://www.omegat.org/

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/
package tokyo.northside.omegat.theme;

import com.formdev.flatlaf.FlatLightLaf;
import org.omegat.gui.theme.DefaultFlatTheme;

import java.awt.Color;
import java.awt.Insets;

import javax.swing.UIDefaults;
import javax.swing.UIManager;
import javax.swing.border.MatteBorder;

/**
 * Flat and round light theme for OmegaT.
 * This is based on FlatLaf's FlatLightLaf module.
 */
@SuppressWarnings("unused")
public class FlatRoundLightLaf extends FlatLightLaf {
    private static final String NAME = "Flat round light theme";
    private static final String ID = "FlatRoundLightTheme";
    private static final String DESCRIPTION = "Rounded theme customized from FlatLightLaf";

    /**
     * Constructor.
     */
    public FlatRoundLightLaf() {
        super();
    }

    /**
     * Return default theme configurations.
     *
     * @return
     */
    @Override
    public UIDefaults getDefaults() {
        UIDefaults defaults = super.getDefaults();
        defaults = DefaultFlatTheme.setDefaults(defaults, ID);

        Color standardBgColor = defaults.getColor("Panel.background");
        defaults.put("TextPane.background", Color.WHITE);

        // Borders
        Color borderColor = new Color(0x0f5e9c);
        defaults.put("OmegaTBorder.color", borderColor);
        defaults.put("OmegaTDockablePanel.border", new MatteBorder(1, 1, 1, 1, borderColor));

        // OmegaT-defined Dockables.
        defaults.put("OmegaTDockablePanel.border", new MatteBorder(1, 1, 1, 1, borderColor));
        defaults.put("OmegaTDockablePanel.isProportionalMargins", true);

        // Default application Colors
        defaults.put("OmegaT.activeSource", new Color(0xc0ffc0));
        defaults.put("OmegaT.source", new Color(0xc0ffc0));
        defaults.put("OmegaT.noted", new Color(0xc0ffff));
        defaults.put("OmegaT.untranslated", new Color(0xc0c0ff));
        defaults.put("OmegaT.translated", new Color(0xffff99));
        defaults.put("OmegaT.nonUnique", new Color(0x808080));
        defaults.put("OmegaT.placeholder", new Color(0x969696));
        defaults.put("OmegaT.removeTextTarget", new Color(0xff0000));
        defaults.put("OmegaT.nbsp", new Color(0xc8c8c8));
        defaults.put("OmegaT.whiteSpace", new Color(0x808080));
        defaults.put("OmegaT.bidiMarkers", new Color(0xc80000));
        defaults.put("OmegaT.paragraphStart", new Color(0xaeaeae));
        defaults.put("OmegaT.markComesFromTm", new Color(0xfa8072));
        defaults.put("OmegaT.markComesFromTxXice", new Color(0xaf76df));
        defaults.put("OmegaT.markComesFromTmX100pc", new Color(0xff9408));
        defaults.put("OmegaT.markComesFromTmXauto", new Color(0xff9596));
        defaults.put("OmegaT.markComesFromTmXendorced", new Color(0xffccff));
        defaults.put("OmegaT.replace", new Color(0x0000ff));
        defaults.put("OmegaT.languageTools", new Color(0x0000ff));
        defaults.put("OmegaT.transTips", new Color(0x0000ff));
        defaults.put("OmegaT.spellCheck", new Color(0xff0000));
        defaults.put("OmegaT.terminology", new Color(0xff6600));
        defaults.put("OmegaT.matchesChanged", new Color(0x0000ff));
        defaults.put("OmegaT.matchesUnchanged", new Color(0x00ff00));
        defaults.put("OmegaT.matchesInsActive", new Color(0x0000ff));
        defaults.put("OmegaT.matchesInsInactive", new Color(0x808080));
        defaults.put("OmegaT.hyperlink", new Color(0x0000ff));
        defaults.put("OmegaT.searchFoundMark", new Color(0x0000ff));
        defaults.put("OmegaT.searchReplaceMark", new Color(0xff9900));
        defaults.put("OmegaT.notificationMin", new Color(0xfff2d4));
        defaults.put("OmegaT.notificationMax", new Color(0xff9900));
        defaults.put("OmegaT.alignerAccepted", new Color(0x15bb45));
        defaults.put("OmegaT.alignerNeedsReview", new Color(0xff0000));
        defaults.put("OmegaT.alignerHighlight", new Color(0xffff00));
        defaults.put("OmegaT.alignerTableRowHighlight", new Color(0xc8c8c8));
        defaults.put("OmegaT.projectFilesCurrentFileForeground", new Color(0x0));
        defaults.put("OmegaT.projectFilesCurrentFileBackground", new Color(0xc8ddf2));
        defaults.put("OmegaT.searchFieldErrorText", new Color(0xff0000));
        defaults.put("OmegaT.searchDimmedBackground", new Color(0x80, 0x80, 0x80, 0x80));
        defaults.put("OmegaT.searchResultBorder", new Color(0xEE, 0xD2, 0x00, 0x80));
        defaults.put("OmegaT.machinetranslateSelectedHighlight", new Color(0xffff00));

        // FIXME: VLDocking values have to be set to the "developer defaults"
        // not the "LAF defaults" because that's where
        // DockingUISettings#installUI puts them
        UIManager.put("DockViewTitleBar.border", new MatteBorder(1, 1, 1, 1, borderColor));

        int cornerRadius = 0;

        // Panel title bars
        Color activeTitleText = defaults.getColor("Label.foreground");
        Color activeTitleBgColor = Utils.adjustRGB(standardBgColor, 0xF6 - 0xEE);
        Color inactiveTitleText = new Color(0x808080);
        defaults.put("InternalFrame.activeTitleForeground", activeTitleText);
        defaults.put("InternalFrame.activeTitleBackground", activeTitleBgColor);
        defaults.put("InternalFrame.inactiveTitleForeground", inactiveTitleText);
        defaults.put("InternalFrame.inactiveTitleBackground", standardBgColor);

        // Undocked panel
        defaults.put("activeCaption", Color.WHITE);
        defaults.put("activeCaptionBorder", borderColor);
        defaults.put("inactiveCaption", standardBgColor);
        defaults.put("inactiveCaptionBorder", borderColor);

        // Standard components customize
        defaults.put("Component.arrowType", "triangle");
        defaults.put("Button.arc", 999);
        defaults.put("Component.arc", 999);
        defaults.put("CheckBox.arc", 999);
        defaults.put("ProgressBar.arc", 999);
        defaults.put("TextComponent.arc", 999);
        defaults.put("ScrollBar.trackArc", 999);
        defaults.put("ScrollBar.thumbArc", 999);
        defaults.put("ScrollBar.trackInsets", new Insets(2, 4, 2, 4));
        defaults.put("ScrollBar.thumbInsets", new Insets(2, 2, 2, 2));
        defaults.put("ScrollBar.track", new Color(0xe0e0e0));
        defaults.put("TabbedPane.tabSeparatorsFullHeight", true);
        defaults.put("TabbedPane.selectedBackground", Color.white);

        return defaults;
    }

    /**
     * Adds this look and feel to the set of available look and feels.
     */
    public static void installLafInfo() {
        installLafInfo(NAME, FlatRoundLightLaf.class);
    }

    @Override
    public String getID() {
        return ID;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String getDescription() {
        return DESCRIPTION;
    }

}


src/main/java/tokyo/northside/omegat/theme/FlatRoundDarkTheme.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2021 Hiroshi Miura
               Home page: http://www.omegat.org/

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/
package tokyo.northside.omegat.theme;

import com.formdev.flatlaf.FlatDarkLaf;

import javax.swing.*;

public class FlatRoundDarkTheme extends FlatDarkLaf {
    private static final String NAME = "Flat round dark theme";
    private static final String ID = "FlatRoundDarkTheme";
    private static final String DESCRIPTION = "Rounded theme customized from FlatDarkLaf";

    public FlatRoundDarkTheme() {
        super();
    }

    @Override
	public UIDefaults getDefaults() {
		UIDefaults defaults = super.getDefaults();
        defaults.put( "Component.arrowType", "triangle" );
        defaults.put( "Button.arc", 999 );
        defaults.put( "Component.arc", 999 );
        defaults.put( "ProgressBar.arc", 999 );
        defaults.put( "TextComponent.arc", 999 );
        return defaults;
    }

  	/**
	 * Adds this look and feel to the set of available look and feels.
	 * <p>
	 * Useful if your application uses {@link UIManager#getInstalledLookAndFeels()}
	 * to query available LaFs and display them to the user in a combobox.
	 */
	public static void installLafInfo() {
		installLafInfo( NAME, FlatRoundDarkTheme.class );
	}

    @Override
    public String getID() {
        return ID;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String getDescription() {
        return DESCRIPTION;
    }
}



src/main/java/tokyo/northside/omegat/theme/RoundedCornerBorder.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2015 Aaron Madlon-Kay
               Home page: http://www.omegat.org/
               Support center: https://omegat.org/support

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/
package tokyo.northside.omegat.theme;

import javax.swing.border.AbstractBorder;
import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Component;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Insets;
import java.awt.RenderingHints;
import java.awt.Shape;
import java.awt.Stroke;
import java.awt.geom.Area;
import java.awt.geom.Line2D;
import java.awt.geom.Rectangle2D;
import java.awt.geom.RoundRectangle2D;


/**
 * @author Aaron Madlon-Kay
 */
@SuppressWarnings("serial")
public class RoundedCornerBorder extends AbstractBorder {

    public static final int SIDE_TOP = 0;
    public static final int SIDE_LEFT = 1;
    public static final int SIDE_BOTTOM = 2;
    public static final int SIDE_RIGHT = 3;
    public static final int SIDE_ALL = 4;

    private final int radius;
    private final Color color;
    private final int side;
    private final int stroke;
    private final transient Stroke strokeObj;

    public RoundedCornerBorder() {
        this(-1, Color.GRAY, SIDE_ALL);
    }

    public RoundedCornerBorder(int radius, Color color, int side) {
        this(radius, color, side, 1);
    }

    public RoundedCornerBorder(int radius, Color color, int side, int stroke) {
        this.radius = radius;
        this.color = color;
        this.side = side;
        this.stroke = stroke;
        this.strokeObj = new BasicStroke(stroke);
    }

    @Override
    public void paintBorder(Component c, Graphics g, int x, int y, int width, int height) {
        Graphics2D g2 = (Graphics2D) g.create();
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        g2.setStroke(strokeObj);

        int r = radius == -1 ? height - stroke : radius;
        RoundRectangle2D roundRect = new RoundRectangle2D.Float(x, y, width - (stroke - 0.5f), height - stroke, r, r);
        Rectangle2D sharpRect = new Rectangle2D.Float(x, y, width, height);
        Area corners = new Area(sharpRect);
        corners.subtract(new Area(roundRect));
        Color background = c.getParent() == null ? null : c.getParent().getBackground();

        if (side == SIDE_ALL) {
            drawCorners(g2, background, corners, roundRect);
            g2.dispose();
            return;
        }

        Shape initialClip = g2.getClip();
        Rectangle2D roundedHalfClip;
        Shape line1;
        Shape line2;
        if (side == SIDE_TOP) {
            roundedHalfClip = new Rectangle2D.Float(x, y, width, height / 2);
            line1 = new Line2D.Float(x, y, x, y + height);
            line2 = new Line2D.Float(x + width - 0.5f, y, x + width - 0.5f, y + height);
        } else if (side == SIDE_LEFT) {
            roundedHalfClip = new Rectangle2D.Float(x, y, width / 2 + stroke, height);
            line1 = new Line2D.Float(x, y, x + width, y);
            line2 = new Line2D.Float(x, y + height - stroke, x + width, y + height - stroke);
        } else if (side == SIDE_BOTTOM) {
            roundedHalfClip = new Rectangle2D.Float(x, y + height / 2, width, height / 2 + stroke);
            line1 = new Line2D.Float(x, y, x, y + height);
            line2 = new Line2D.Float(x + width - (stroke - 0.5f), y, x + width - (stroke - 0.5f), y + height);
        } else if (side == SIDE_RIGHT) {
            roundedHalfClip = new Rectangle2D.Float(x + width / 2, y, width / 2 + stroke, height);
            line1 = new Line2D.Float(x, y, x + width, y);
            line2 = new Line2D.Float(x, y + height - stroke, x + width, y + height - stroke);
        } else {
            throw new IllegalArgumentException();
        }
        g2.clip(roundedHalfClip);
        drawCorners(g2, background, corners, roundRect);

        Area inverseClip = new Area(sharpRect);
        inverseClip.subtract(new Area(roundedHalfClip));
        g2.setClip(initialClip);
        g2.clip(inverseClip);
        g2.draw(line1);
        g2.draw(line2);

        g2.dispose();
    }

    private void drawCorners(Graphics2D g2, Color background, Area corners, Shape shape) {
        if (background != null) {
            g2.setColor(background);
            g2.fill(corners);
        }
        g2.setColor(color);
        g2.draw(shape);
    }

    @Override
    public Insets getBorderInsets(Component c) {
        return new Insets(4, 8, 4, 8);
    }

    @Override
    public Insets getBorderInsets(Component c, Insets insets) {
        insets.set(4, 8, 4, 8);
        return insets;
    }
}


src/main/java/tokyo/northside/omegat/theme/FlatRoundDarkLaf.java
/**************************************************************************
 OmegaT - Computer Assisted Translation (CAT) tool
          with fuzzy matching, translation memory, keyword search,
          glossaries, and translation leveraging into updated projects.

 Copyright (C) 2021 Hiroshi Miura
               Home page: http://www.omegat.org/

 This file is part of OmegaT.

 OmegaT is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 OmegaT is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **************************************************************************/
package tokyo.northside.omegat.theme;

import com.formdev.flatlaf.FlatDarkLaf;
import org.omegat.gui.theme.DefaultFlatTheme;

import javax.swing.UIDefaults;
import javax.swing.UIManager;
import javax.swing.border.MatteBorder;
import java.awt.Color;
import java.awt.Insets;


/**
 * Flat and round dark theme for OmegaT.
 * This is based on FlatLaf's FlatDarkLaf module.
 */
@SuppressWarnings("unused")
public class FlatRoundDarkLaf extends FlatDarkLaf {
    private static final String NAME = "Flat round dark theme";
    private static final String ID = "FlatRoundDarkTheme";
    private static final String DESCRIPTION = "Rounded theme customized from FlatDarkLaf";

    /**
     * Constructor.
     */
    public FlatRoundDarkLaf() {
        super();
    }

    /**
     * Return default theme configurations.
     *
     * @return
     */
    @Override
    public UIDefaults getDefaults() {
        UIDefaults defaults = super.getDefaults();

        // get OmegaT defaults
        defaults = DefaultFlatTheme.setDefaults(defaults, ID);

        // GTK+ has bug that TextPane background is fixed white.
        // https://sourceforge.net/p/omegat/bugs/1013/
        Color standardBgColor = defaults.getColor("Panel.background");
        defaults.put("TextPane.background", standardBgColor);

        // Borders
        Color borderColor = new Color(0x8aa5b8);
        defaults.put("OmegaTBorder.color", borderColor);
        defaults.put("OmegaTDockablePanel.border", new MatteBorder(1, 1, 1, 1, borderColor));

        // Default application Colors
        defaults.put("OmegaT.activeSource", new Color(0x287128));
        defaults.put("OmegaT.source", new Color(0x82b682));
        defaults.put("OmegaT.noted", new Color(0xa0c0c0));
        defaults.put("OmegaT.untranslated", new Color(0x4d4daa));
        defaults.put("OmegaT.translated", new Color(0x57572d));
        defaults.put("OmegaT.nonUnique", new Color(0x808080));
        defaults.put("OmegaT.placeholder", new Color(0x969696));
        defaults.put("OmegaT.removeTextTarget", new Color(0xff0000));
        defaults.put("OmegaT.nbsp", new Color(0xc8c8c8));
        defaults.put("OmegaT.whiteSpace", new Color(0x808080));
        defaults.put("OmegaT.bidiMarkers", new Color(0x480000));
        defaults.put("OmegaT.paragraphStart", new Color(0xaeaeae));
        defaults.put("OmegaT.markComesFromTm", new Color(0xaa8072));
        defaults.put("OmegaT.markComesFromTxXice", new Color(0xaf76df));
        defaults.put("OmegaT.markComesFromTmX100pc", new Color(0x3f54f8));
        defaults.put("OmegaT.markComesFromTmXauto", new Color(0x3f2546));
        defaults.put("OmegaT.markComesFromTmXendorced", new Color(0x005500));
        defaults.put("OmegaT.replace", new Color(0x0000ff));
        defaults.put("OmegaT.languageTools", new Color(0x0000ff));
        defaults.put("OmegaT.transTips", new Color(0x0000ff));
        defaults.put("OmegaT.spellCheck", new Color(0xff0000));
        defaults.put("OmegaT.terminology", new Color(0xffa500));
        defaults.put("OmegaT.matchesChanged", new Color(0x0000ff));
        defaults.put("OmegaT.matchesUnchanged", new Color(0x00ff00));
        defaults.put("OmegaT.matchesInsActive", new Color(0x0000ff));
        defaults.put("OmegaT.matchesInsInactive", new Color(0x808080));
        defaults.put("OmegaT.hyperlink", new Color(0x0000ff));
        defaults.put("OmegaT.searchFoundMark", new Color(0x0000ff));
        defaults.put("OmegaT.searchReplaceMark", new Color(0xff9900));
        defaults.put("OmegaT.notificationMin", new Color(0x332233));
        defaults.put("OmegaT.notificationMax", new Color(0x647354));
        defaults.put("OmegaT.alignerAccepted", new Color(0x15bb45));
        defaults.put("OmegaT.alignerNeedsReview", new Color(0xff0000));
        defaults.put("OmegaT.alignerHighlight", new Color(0xffff00));
        defaults.put("OmegaT.alignerTableRowHighlight", new Color(0xc8c8c8));
        defaults.put("OmegaT.projectFilesCurrentFileForeground", new Color(0x0));
        defaults.put("OmegaT.projectFilesCurrentFileBackground", new Color(0xc8ddf2));
        defaults.put("OmegaT.searchFieldErrorText", new Color(0x7f0000));
        defaults.put("OmegaT.searchDimmedBackground", new Color(0x80, 0x80, 0x80, 0x80));
        defaults.put("OmegaT.searchResultBorder", new Color(0xEE, 0xD2, 0x00, 0x80));
        defaults.put("OmegaT.machinetranslateSelectedHighlight", new Color(0xff9900));

        // FIXME: VLDocking values have to be set to the "developer defaults"
        // not the "LAF defaults" because that's where
        // DockingUISettings#installUI puts them
        UIManager.put("DockViewTitleBar.border", new MatteBorder(1, 1, 1, 1, borderColor));

        // Panel title bars
        Color activeTitleBgColor = Utils.adjustRGB(standardBgColor, 0xF6 - 0xEE);
        Color activeTitleText = defaults.getColor("Label.foreground");
        Color inactiveTitleText = new Color(0x767676);

        defaults.put("InternalFrame.activeTitleForeground", activeTitleText);
        defaults.put("InternalFrame.activeTitleBackground", activeTitleBgColor);
        defaults.put("InternalFrame.inactiveTitleForeground", inactiveTitleText);
        defaults.put("InternalFrame.inactiveTitleBackground", standardBgColor);

        // Undocked panel
        defaults.put("activeCaption", Color.GRAY);
        defaults.put("activeCaptionBorder", borderColor);
        defaults.put("inactiveCaption", standardBgColor);
        defaults.put("inactiveCaptionBorder", borderColor);

        // Standard components customize
        defaults.put("Component.arrowType", "triangle");
        defaults.put("Button.arc", 999);
        defaults.put("Component.arc", 999);
        defaults.put("CheckBox.arc", 999);
        defaults.put("ProgressBar.arc", 999);
        defaults.put("TextComponent.arc", 999);
        defaults.put("ScrollBar.trackArc", 999);
        defaults.put("ScrollBar.thumbArc", 999);
        defaults.put("ScrollBar.trackInsets", new Insets(3, 5, 3, 5));
        defaults.put("ScrollBar.thumbInsets", new Insets(3, 3, 3, 3));
        defaults.put("ScrollBar.width", 16);
        defaults.put("ScrollBar.track", new Color(0x757575));
        defaults.put("TabbedPane.tabSeparatorsFullHeight", true);
        defaults.put("TabbedPane.selectedBackground", Color.GRAY);

        return defaults;
    }

    /**
     * Adds this look and feel to the set of available look and feels.
     * <p>
     * Useful if your application uses {@link UIManager#getInstalledLookAndFeels()}
     * to query available LaFs and display them to the user in a combobox.
     */
    public static void installLafInfo() {
        installLafInfo(NAME, FlatRoundDarkLaf.class);
    }

    /**
     * Return theme id.
     *
     * @return
     */
    @Override
    public String getID() {
        return ID;
    }

    /**
     * Return human readable name of theme.
     *
     * @return
     */
    @Override
    public String getName() {
        return NAME;
    }

    /**
     * Return description of theme.
     *
     * @return
     */
    @Override
    public String getDescription() {
        return DESCRIPTION;
    }
}
